### ISSUE TITLE
Jasper demo crashes out  PCSX2

### ISSUE THREAD

**Issue by benderscruffy** [PCSX2/pcsx2#1302]
PCSX2 version: GIT v1.5.0-dev-625-g87867be 	
PCSX2 options: Defaults, 
Plugins used: GSdx, SPU2-X, Lilypad, Gigaherz CDVD,rest null plugins

Plugin settings: Defaults, 

here is the source 
http://lukasz.dk/files/jaspersrc.zip

after the first credits when it goes to start the demo the emulator crashes out

just google Jasper PS2 Demo (PD).

**Comment by willkuer** [comment]
Tried playing with ee clamping/rounding modes and ee timing hack?

**Comment by Catarax** [comment]
More info about the crash :
![3l5qopb](https://cloud.githubusercontent.com/assets/3964507/14829398/6bcd1f6e-0beb-11e6-8018-3aa787e405f7.png)

**Comment by ssakash** [comment]
I  rather doubt that it's related to the crash. Does it crash instantly after the SYSCALL message ?

**Comment by MrCK1** [comment]
Took a quick look at this, don't get the GS messages. The crash is related to the GIF Unit. Could be useful for debugging games in the future; so here's half of the callstack.

```
c:\users\chris\documents\github\pcsx2\pcsx2\Gif_Unit.h(263) : assertion failed:
    Function:  Gif_Path::CopyGSPacketData
    Thread:    EE Core
    Condition: curSize+size<=buffSize
    Message:   Gif Path Buffer Overflow!

[00] pxOnAssert                                   \documents\github\pcsx2\common\src\utilities\exceptions.cpp:134
[01] pxOnAssert                                   \documents\github\pcsx2\common\src\utilities\exceptions.cpp:129
[02] Gif_Path::CopyGSPacketData                   \documents\github\pcsx2\pcsx2\gif_unit.h:263
[03] Gif_Unit::TransferGSPacketData               \documents\github\pcsx2\pcsx2\gif_unit.h:526
[04] GIF_Fifo::read                               \documents\github\pcsx2\pcsx2\gif.cpp:124
[05] gifInterrupt                                 \documents\github\pcsx2\pcsx2\gif.cpp:197
[06] TESTINT                                      \documents\github\pcsx2\pcsx2\r5900.cpp:264
[07] _cpuTestInterrupts                           \documents\github\pcsx2\pcsx2\r5900.cpp:283
[08] _cpuEventTest_Shared                         \documents\github\pcsx2\pcsx2\r5900.cpp:399
[09] recEventTest                                 \documents\github\pcsx2\pcsx2\x86\ix86-32\ir5900-32.cpp:338
[10] eeRecDispatchers                            
[11] recExecute                                   \documents\github\pcsx2\pcsx2\x86\ix86-32\ir5900-32.cpp:690
```

**Comment by F0bes** [comment]
Just debugged an issue with a library in the SDK that does this exact thing.
It ends up sending a GIFTag (or complete junk) with a crazy high NLOOP, making the GIF expect a lot more than it's going to receive. Because of this, I believe the GIF is holding every transfer that happens in it's buffer, with the intent to kick it to the GS once it received the insane amount of QW it was accidentally told it would receive.

I'm not familiar enough with this code, but the suspect might be the packet realignment, going off of the log right before the crash
```
Swapping to PATH 3
PATH 3 New tag State 3 FLG 1 EOP 0 NLOOP 1
PATH 3 not enough data, available 1926784 wanted 4889360
GIF_TRANS_DMA - [path=3][size=524176]
Swapping to PATH 3
PATH 3 New tag State 3 FLG 1 EOP 0 NLOOP 1
PATH 3 not enough data, available 2450960 wanted 4889360
GIF_TRANS_DMA - [path=3][size=128]
Swapping to PATH 3
PATH 3 New tag State 3 FLG 1 EOP 0 NLOOP 1
PATH 3 not enough data, available 2451088 wanted 4889360
GIF_TRANS_DMA - [path=3][size=96]
Swapping to PATH 3
PATH 3 New tag State 3 FLG 1 EOP 0 NLOOP 1
PATH 3 not enough data, available 2451184 wanted 4889360
GIF_TRANS_DMA - [path=3][size=307072]
CopyGSPacketData: Realigning packet!
Path Buffer: Realigning packet!
Realign Packet [9416960]
/home/ty/pcsx2/pcsx2/./Gif_Unit.h:322: assertion failed in function void Gif_Path::CopyGSPacketData(u8 *, u32, bool): Gif Path Buffer Overflow!
```

Here is also an updated call stack

```
AbortWithMessage(const char * msg) (/home/ty/pcsx2/common/Misc.cpp:100)
pxOnAssertFail(const char * file, int line, const char * func, const char * msg) (/home/ty/pcsx2/common/Exceptions.cpp:141)
Gif_Path::CopyGSPacketData(Gif_Path * this, u8 * pMem, u32 size, bool aligned) (/home/ty/pcsx2/pcsx2/Gif_Unit.h:322)
Gif_Unit::TransferGSPacketData(Gif_Unit * this, GIF_TRANSFER_TYPE tranType, u8 * pMem, u32 size, bool aligned) (/home/ty/pcsx2/pcsx2/Gif_Unit.h:683)
WRITERING_DMA(u32 * pMem, u32 qwc) (/home/ty/pcsx2/pcsx2/Gif.cpp:349)
GIFchain() (/home/ty/pcsx2/pcsx2/Gif.cpp:368)
GIFdma() (/home/ty/pcsx2/pcsx2/Gif.cpp:476)
gifInterrupt() (/home/ty/pcsx2/pcsx2/Gif.cpp:300)
dmaGIF() (/home/ty/pcsx2/pcsx2/Gif.cpp:505)
DmaExec(void (*)(void) func, u32 mem, u32 value) (/home/ty/pcsx2/pcsx2/Dmac.cpp:289)
dmacWrite32<10u>(u32 mem, mem32_t & value) (/home/ty/pcsx2/pcsx2/Dmac.cpp:411)
_hwWrite32<10u>(u32 mem, u32 value) (/home/ty/pcsx2/pcsx2/HwWrite.cpp:139)
hwWrite32<10u>(u32 mem, u32 value) (/home/ty/pcsx2/pcsx2/HwWrite.cpp:281)
m_IndirectDispatchers (Unknown Source:0)
[Unknown/Just-In-Time compiled code] (Unknown Source:0)
recExecute() (/home/ty/pcsx2/pcsx2/x86/ix86-32/iR5900-32.cpp:763)
```

**Comment by refractionpcsx2** [comment]
What the hell, how does nothing else get affected by this??? also wtf.

**Comment by Goatman13** [comment]
Yeah. I complete don't understand how that happen. I could understand if whole elf get loaded with offset, or different segment, etc. But this just happen at "random" point in elf with single segment.

**Comment by refractionpcsx2** [comment]
If we weren't manually loading it, I'd say maybe it's a SIF/IOP bug somewhere, but this is a homebrew ELF, which means we're directly injecting it in to EE memory, how the heck is it getting offset? :/

**Comment by refractionpcsx2** [comment]
Kind of a side note, but I also tried this on Dobiestation and it crashed at the same point as PCSX2, also with a GIF packet problem.

**Comment by refractionpcsx2** [comment]
So I dumped the memory on my PS2 when the demo was running, and (with the 0x100000 offset) I checked, and the memory matches PCSX2..

![image](https://user-images.githubusercontent.com/6278726/210918169-30180135-a650-40e3-ac45-50edbb7c323e.png)

**Comment by Goatman13** [comment]
Interesting. Well, hardware test show that is correct, then it is correct. I loaded that elf into Play! emulator just see how that one handle it. Result is opposite to pcsx2/real hw, and match ps2dis/IDA/Ghidra. Demo don't even reach point where it hang on other emulators.

![image](https://user-images.githubusercontent.com/101417270/210972713-8532e42f-6a51-4f18-add8-4777c54103ce.png)

This is not really related to the issue now, but what kind of weird elf file is that? Difference always starts at the same point, and absolute jumps to address above 0x410350 seems to always point to wrong address on real HW/PCSX2.

**Comment by refractionpcsx2** [comment]
Yeah I have no clue why that's happening! I guess there must be some sort of self modification going on, I mean this area seems to be some sort of pointers or state information? I dunno.... I'm not really familiar with the Jasper source.

Maybe you should look for the Jasper source, I believe it's out there, Koz you seem to have an idea what you're doing :)

**Comment by seta-san** [comment]
i believe i found the source in case anyone wants it.
[jaspersrc.zip](https://github.com/PCSX2/pcsx2/files/10471759/jaspersrc.zip)

**Comment by refractionpcsx2** [comment]
Certainly looks like it is (I didn't compile this, just comparing text to my already compiled version)
![image](https://user-images.githubusercontent.com/6278726/213868425-52afb81a-3317-4233-95e8-2cb1d4e7a31d.png)

**Comment by seta-san** [comment]
still a problem. 2.9.9

**Comment by Goatman13** [comment]
> Related to this & also by the same developer; 3StarsII homebrew demo black screens with TLB Misses

This is different issue. Interesting one, but different. Demo creates new thread initial_priority using current ThreadParam data, it substrat 1 from current thread current_priority and then create new thread with it as initial_priority.

Problem is that current thread is running with 0 priority, making new thread initial_priority 0xFFFFFFFFFFFFFFFF. This later causes oob read for priority table and from here everything goes wrong as it is executing random stuff. 

Problem is that I don't understand how that works on real PS2. This is something that pcsx2 interfere with by itself but shouldn't affect it anyway (during debug thread list reading on create thread syscall). The rest of ThreadParam data looks ok, isn't malformed by some ee cache stuff or something like that. It looks like same thing should happen on PS2, but it is not.
Just in case you want to see this demo anyway, on 004007B4 li a3, 1 should so the job.

**Comment by Ziemas** [comment]
When I last looked at jasper I tried debugging it on my PS2 with rdb/dsedb, but I couldn't even get it to work on a real PS2 because it doesn't set up its own stack properly. IIRC I ended up trying to write my own elf loader to work around that but I think I failed to come to any useful conclusions.

Regarding 3starsii I came to the same conclusion as Goatman a while back, I think they must have had a homebrew loader at the time that set up thread stuff before running the elf, because IIRC it can't work as-is when loaded by the kernel.

**Comment by Goatman13** [comment]
> Regarding 3starsii I came to the same conclusion as Goatman a while back, I think they must have had a homebrew loader at the time that set up thread stuff before running the elf, because IIRC it can't work as-is when loaded by the kernel

You're right 3starsii work fine when launched from old LaunchELFv3.41a. I'm pretty sure that Jasper demo will work on some specific elf loader they used back in the days. Just in case, it won't work with that LaunchELF.
